enclave/src/channel_storage.rs

```diff
@@ -0,0 +1,96 @@
+// This file is part of Polkadex.
+
+// Copyright (C) 2020-2021 Polkadex oü and Supercomputing Systems AG
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::polkadex_balance_storage::{Balances, PolkadexBalanceKey};
+use crate::polkadex_gateway::GatewayError;
+use log::*;
+use polkadex_sgx_primitives::AccountId;
+use sp_std::prelude::*;
+use std::sync::{
+    atomic::{AtomicPtr, Ordering},
+    mpsc::{channel, Receiver, Sender},
+    Arc, SgxMutex, SgxMutexGuard,
+};
+
+static GLOBAL_CHANNEL_STORAGE: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+
+pub fn create_in_memory_channel_storage() -> Result<(), GatewayError> {
+    let storage = ChannelStorage::default();
+    let storage_ptr = Arc::new(SgxMutex::<ChannelStorage>::new(storage));
+    let ptr = Arc::into_raw(storage_ptr);
+    GLOBAL_CHANNEL_STORAGE.store(ptr as *mut (), Ordering::SeqCst);
+    Ok(())
+}
+
+pub struct ChannelStorage {
+    pub sender: Sender<ChannelType>,
+    pub receiver: Arc<SgxMutex<Receiver<ChannelType>>>,
+}
+
+impl Default for ChannelStorage {
+    fn default() -> Self {
+        let (sender, receiver) = channel();
+        Self {
+            sender,
+            receiver: Arc::new(SgxMutex::new(receiver)),
+        }
+    }
+}
+
+pub enum ChannelType {
+    Nonce(AccountId, u32),
+    Balances(PolkadexBalanceKey, Balances),
+}
+
+pub fn load_sender() -> Result<Sender<ChannelType>, ChannelStorageError> {
+    // Acquire lock on proxy_registry
+    let mutex = load_channel_storage()?;
+    let storage: SgxMutexGuard<ChannelStorage> = mutex
+        .lock()
+        .map_err(|_| ChannelStorageError::CouldNotGetMutex)?;
+    let result = storage.sender.clone();
+    Ok(result)
+}
+
+pub fn load_receiver() -> Result<Arc<SgxMutex<Receiver<ChannelType>>>, ChannelStorageError> {
```

Comment:
I'm using Arc so i can use .clone()

Reply:
But why is clone necessary for the receiver?

Reply:
Because it is not being used right there, unless there's another solution to save it for later use when calling the function without having to clone it.

Reply:
Removed the need for the clone by scraping the load_receover() function and directly loading it in lib.rs